controllers/k6_start.go

```diff
@@ -3,105 +3,140 @@ package controllers
 import (
 	"context"
 	"fmt"
-	"net/http"
-	"time"
-
 	"github.com/go-logr/logr"
 	"github.com/grafana/k6-operator/api/v1alpha1"
 	"github.com/grafana/k6-operator/pkg/resources/jobs"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/labels"
-	"k8s.io/apimachinery/pkg/util/wait"
+	"net/http"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
 )
 
+var backoffSchedule = []time.Duration{
+	1 * time.Second,
+	3 * time.Second,
+	5 * time.Second,
+}
+
 func isServiceReady(log logr.Logger, service *v1.Service) bool {
-	resp, err := http.Get(fmt.Sprintf("http://%v.%v.svc.cluster.local:6565/v1/status", service.ObjectMeta.Name, service.ObjectMeta.Namespace))
+	var resp *http.Response
+	var err error
+	for _, backoff := range backoffSchedule {
+		resp, err = http.Get(fmt.Sprintf("http://%v.%v.svc.cluster.local:6565/v1/status", service.ObjectMeta.Name, service.ObjectMeta.Namespace))
+
+		if err == nil && resp.StatusCode < 299 {
+			break
+		}
+
+		time.Sleep(backoff)
+	}
```

Comment:
This is related with your comment about the  `liveness probe`. Sometimes even the runner is ready by the K8s point of view this was not true. That's the main reason.

I can remove it but in this case, I think we should include also a liveness and readiness probe or most of the times the starter job will never run :(

Reply:
Update about this question. Removing the pulling causes some errors about service state.

Logs in the operator:

```
2023-01-20T17:34:41.950Z        ERROR   controllers.K6  failed to get status from k6-sample-service-3   {"namespace": "default", "name": "k6-sample", "error": "Get \"http://k6-sample-service-3.default.svc.cluster.local:6565/v1/status\": dial tcp 10.101.56.166:6565: connect: connection refused"}

[Removed stack trace for clarity]

2023-01-20T17:34:41.951Z        INFO    controllers.K6  k6-sample-service-3 service is not ready, aborting      {"namespace": "default", "name": "k6-sample"}
2023-01-20T17:34:41.951Z        DEBUG   controller      Successfully Reconciled {"reconcilerGroup": "k6.io", "reconcilerKind": "K6", "controller": "k6", "name": "k6-sample", "namespace": "default"}
2023-01-20T17:36:28.861Z        INFO    controllers.K6  Reconcile(); stage = created    {"namespace": "default", "name": "k6-sample"}
2023-01-20T17:36:28.861Z        INFO    controllers.K6  Waiting for pods to get ready   {"namespace": "default", "name": "k6-sample"}
2023-01-20T17:36:28.861Z        INFO    controllers.K6  3/3 runner pods ready   {"namespace": "default", "name": "k6-sample"}
2023-01-20T17:36:28.872Z        INFO    controllers.K6  k6-sample-service-2 service is ready    {"namespace": "default", "name": "k6-sample"}
2023-01-20T17:36:28.888Z        INFO    controllers.K6  k6-sample-service-3 service is ready    {"namespace": "default", "name": "k6-sample"}
2023-01-20T17:36:28.896Z        INFO    controllers.K6  k6-sample-service-1 service is ready    {"namespace": "default", "name": "k6-sample"}
2023-01-20T17:36:28.896Z        INFO    controllers.K6  Changing stage of K6 status to started  {"namespace": "default", "name": "k6-sample"}

```

K6 Pod logs:

```

time="2023-01-20T17:34:33Z" level=debug msg="Initialized VU #198" phase=local-execution-scheduler-init                                                                
time="2023-01-20T17:34:33Z" level=debug msg="Initialized VU #186" phase=local-execution-scheduler-init                                                                time="2023-01-20T17:34:33Z" level=debug msg="Initialized VU #195" phase=local-execution-scheduler-init                                                                time="2023-01-20T17:34:33Z" level=debug msg="Finished initializing needed VUs, start initializing executors..." phase=local-execution-scheduler-init
time="2023-01-20T17:34:33Z" level=debug msg="Initialized executor default" phase=local-execution-scheduler-init
time="2023-01-20T17:34:33Z" level=debug msg="Initialization completed" phase=local-execution-scheduler-init                                                           time="2023-01-20T17:34:33Z" level=debug msg="Execution scheduler starting..." component=engine                                                                        
time="2023-01-20T17:34:33Z" level=debug msg="Execution is paused, waiting for resume or interrupt..." phase=local-execution-scheduler-run
time="2023-01-20T17:34:33Z" level=debug msg="Metrics processing started..." component=engine                                                                          time="2023-01-20T17:34:40Z" level=debug msg="GET /v1/status" status=200                                                                                               time="2023-01-20T17:34:40Z" level=debug msg="GET /v1/status" status=200
```

It seems until the executor is initialized the k6 REST API has not been started ¿this is right?

In this case, I think it could "not so bad" to use the retry strategy in the method `isServiceAvailable()`. By far, as you have mention, the best approach is provide K8s friendly /status endpoint :)

Reply:
:thinking: I think I'm missing something here.

Going by these ^ logs you posted, the service failed once - controller stayed in `created`. Then on the next loop, service was ready and so controller moved to `started`. That's not an erroneous scenario: it should still proceed to execute the test. Did this test run fail to start here?

> It seems until the executor is initialized the k6 REST API has not been started ¿this is right?

No... or at least it shouldn't be so. In runner logs, before the VUs messages (`Initialized VU #198`), you should see something like this:
```
DEBU[0000] Starting the REST API server on localhost:6565 
```
So k6 starts REST pretty early, before kick-starting VUs. And the last message `Metrics processing started` means that it's just waiting for the starter.

IOW, I'm not clear what the error situation here is :sweat_smile:

Reply:
I try to explain a bit better :).

The first thing to clarify is that the operator, including the changes in this PR, **correctly triggers the load test**.

Right now when the runner pods change to `Running` state the operator tries to start the test by sending a request to the `/v1/status` endpoint. But during the k6 initialization stage, some of these requests are failing and for me is not clear the reason why :(.

As the Operator's Reconcile() loop continues to run, as soon as the runner pod responds correctly, the test will start, but during those checks we will see those connection error messages.

The behaviour I expected when configuring the Readiness and Liveness nodes using the `/v1/status`, is to be able to make the start call only when the runner is `fully initialised` and ready to launch the load test.

Thanks!

Reply:
Ivan, I got a repeat of this behaviour: indeed, as you described, it fails to start sometimes. I think it happens because of how events are being timed... We can solve this with the requeue of the event. I'll add a suggestion in the relevant part of the code now - please try it when you get a moment!

Reply:
I think is a good option to solve this point